File: Problem_1.js
// N Queens(https://leetcode.com/problems/n-queens/)

// TC: O(n!)
// SC: O(n^2)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach

let result;
let nLength;
let grid;
var isValid = (row, col) => {
    // Check above cols
    let i = row - 1;
    while (i >= 0) {
        if (grid[i][col] === true)
            return false;
        i--;
    }
    // Check top left diagonals
    i = row - 1;
    let j = col - 1;
    while (i >= 0 && j >= 0) {
        if (grid[i][j] === true)
            return false;
        i--; j--;
    }
    // Check top right diagonals
    i = row - 1;
    j = col + 1;
    while (i >= 0 && col < nLength) {
        if (grid[i][j] === true)
            return false;
        i--; j++;
    }
    return true;
}
var backtrack = (row) => {
    // Base
    if (row === nLength) {
        let outerGrid = [];
        for (let i = 0; i < nLength; i++) {
            let innerGrid = "";
            for (let j = 0; j < nLength; j++) {
                if (grid[i][j]) {
                    innerGrid += 'Q';
                } else {
                    innerGrid += '.';
                }
            }
            outerGrid.push(innerGrid);
        }
        result.push(outerGrid)
        return;
    }
    // Logic
    for (let col = 0; col < nLength; col++) {
        if (isValid(row, col)) {
            // Action
            grid[row][col] = true;
            // Recurse
            backtrack(row + 1);
            // Backtrack
            grid[row][col] = false;
        }
    }
}
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function (n) {
    if (n === 0)
        return [];
    nLength = n;
    result = [];
    grid = new Array(n);
    for (let i = 0; i < n; i++) {
        grid[i] = new Array(n);
        grid[i].fill(false);
    }
    // console.log(grid)
    backtrack(0);

    return result;
};

File: problem_1.java
/*

N Queens(https://leetcode.com/problems/n-queens/)


Time Complexity : O(n!)
Space Complexity : O(n^2)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

*/


class Solution {
    List<List<String>> result;
    boolean[][] grid;
    public List<List<String>> solveNQueens(int n) {
        if(n == 0){
            return new ArrayList<>();
        }

        grid = new boolean[n][n];
        result = new ArrayList<>();
        recurse(0);
        return result;
    }

    private void recurse(int row){

        if(row == grid.length){
            List<String> add = new ArrayList<>();
            for(int i =0; i<grid.length; i++){
                StringBuilder s = new StringBuilder();
                for(int j = 0; j<grid.length; j++){
                    if(grid[i][j]){
                        s.append('Q');
                    }
                    else{
                        s.append('.');
                    }
                }
                add.add(s.toString());
            }
            result.add(add);
            return;
        }

        for(int j = 0; j < grid.length; j++){
            if(isAllowed(row, j)){
                grid[row][j] = true;
                recurse(row + 1);
                grid[row][j] = false;

            }
        }

    }

    private boolean isAllowed(int row, int col){
        int row1 = row;
        int col1 = col;
        while(row1 >= 0){
            if(grid[row1][col1]){
                return false;
            }
           row1--;
        }
        row1 = row;
        while(row1>= 0 && col1>= 0){
            if(grid[row1][col1]){
                return false;
            }
            row1--;
            col1--;
        }
        row1 = row;
        col1 = col;
        while(row1>= 0 && col1< grid.length){
            if(grid[row1][col1]){
                return false;
            }
            row1--;
            col1++;
        }

        return true;
    }
}